src/connectivity.jl

```diff
@@ -219,31 +219,32 @@ julia> strongly_connected_components(g)
 
 function strongly_connected_components end
 # see https://github.com/mauro3/SimpleTraits.jl/issues/47#issuecomment-327880153 for syntax
-@traitfn function strongly_connected_components(g::AG::IsDirected) where {T<:Integer, AG <: AbstractGraph{T}}
+empty_graph_data(type,g::AG) where {T, AG <: AbstractGraph{T}} = Dict{T,type}()
+empty_graph_data(type,g::AG) where {T<:Integer, AG <: AbstractGraph{T}} = zeros(type,nv(g))
+is_unvisited(data::AbstractVector,v::Integer) = iszero(data[v])
+is_unvisited(data::Dict,v) = !haskey(data,v)
+
+@traitfn function strongly_connected_components(g::AG::IsDirected) where {T, AG <: AbstractGraph{T}}
     zero_t = zero(T)
-    one_t = one(T)
     nvg = nv(g)
-    count = one_t
-    
-    
-    index = zeros(T, nvg)         # first time in which vertex is discovered
-    stack = Vector{T}()           # stores vertices which have been discovered and not yet assigned to any component
-    onstack = zeros(Bool, nvg)    # false if a vertex is waiting in the stack to receive a component assignment
-    lowlink = zeros(T, nvg)       # lowest index vertex that it can reach through back edge (index array not vertex id number)
-    parents = zeros(T, nvg)       # parent of every vertex in dfs
+    count = nvg  # (Counting downwards) Visitation order for the branch being explored. Backtracks when we pop an scc.
```

Comment:
```suggestion
    count = Int(nvg)  # (Counting downwards) Visitation order for the branch being explored. Backtracks when we pop a scc.
```
this is a bit subtle, but `nvg` is not always an `Int`, it could also be an `UInt8` for example. Later when you do `count -= 1` it will be converted to an `Int`. This means that the type of `count` changes during execution, leading to less efficient code, especially when that happens inside a loop. You can use `@code_warntype` and run your code with a graph of type `SimpleDiGraph{UInt8}` to see if your code is type unstable.

Reply:
> 
> 
> Hi, thanks for your contribution - It's been a few years since I last looked at Tarjans algorithm, so I will first have to figure out again how it works to do some complete code review. In the meantime, could you add the code you used for benchmarking and the results that you got as a comment to this PR?

I updated the gist with some basic benchmarks.

Reply:
In the end I ended up making rindex, count, and component_count T-valued as well. Partly to save space when T is represented using smaller types, and partly for correctness on machines where Int = Int32 and a graph explicitly needs to be Int64